### PR TITLE
add homework

### PR DESCRIPTION
**Pull request by AliRezaHassanloo2731**
create getReply()

### CODE REVIEW

File: javascript/javascript1/week4/exercise.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  } else if (int === 2) {
+    fibo = 1;
+  } else {
+    fibo = fib(int - 1) + fib(int - 2);
+  }
+  return fibo;
+}
+
+//*RENDER
+// console.log(fib(0));
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/exercise.js]: I tested out the fib function and it doesn't appear to work. I recommend reviewing the files Rasmus shared after class to see how fibonacci can be done

File: javascript/javascript1/week4/exercise.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      str += " Buzz ";
+      str += "\n";
+    } else {
+      str += i;
+    }
+  }
+
+  return str;
+}
+console.log(fizzBuzz(4, 12));
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/exercise.js]: Good work!

File: javascript/javascript1/week4/exercise.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    if (negativeWords.includes(strArray[i])) {
+      score--;
+      negative.push(strArray[i]);
+      console.log(score);
+    }
+  }
+  // myObj = { score: score, negative: negative, positive: positive };
+  return {
+    score,
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/exercise.js]: This works but you could actually simplify it further if you wanted to by the score being:

`score: positiveWords.length - negativeWords.length` when you return it

File: javascript/javascript1/week4/homework.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  firstName += command.split(" ").slice(-1).join();
+  return firstName;
+}
+
+// GREETING AND NAME
+function greetingAnswer(msg) {
+  let answer = "";
+  findFirstName(msg);
+  msg.includes("Hello my name is")
+    ? (answer += "nice to meet you" + " " + firstName)
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/homework.js]: Either you have to save the returned result from `findFirstName()` as the variable `firstName` or call `findFirstName()` directly in the answer string

File: javascript/javascript1/week4/homework.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  answers.push(anwerObj);
+
+  return answer;
+}
+// console.log(greetingAnswer("Hello my name is benjamin")); //nice to meet you benjamin
+
+//ACTIVITIES
+function createActivity(msg) {
+  let answer = "";
+  activity = msg.substring(msg.indexOf(" "), msg.indexOf("to my todo"));
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/homework.js:32]: You could use similar logic as you had above with `.split` here too and grab the contents directly from the index.

`msg.split(" ")[0]` or [1], [2] etc

File: javascript/javascript1/week4/homework.js [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+  } else if (command === "What day is it today?") {
+    answer = getDatOfToday();
+  } else if (command.startsWith("Add")) {
+    answer = createActivity(command);
+  } else if (command === "What is on my todo?") {
+    answer = createActivity(command);
+  } else if (command.includes("Remove")) {
+    answer = removeActivity(command);
+  } else if (command.includes("what is")) {
+    answer = mathematicalCalc(command);
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/homework.js]: It's really nice how you abstracted each of the condition logic away and created functions!

File: javascript/javascript1/week4/homework.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+  let answer = "";
+  activity = msg.substring(msg.indexOf(" "), msg.indexOf("to my todo"));
+
+  if (msg.startsWith("Add")) {
+    activities.push(activity);
+  } else if (msg === "What is on my todo?") {
+    answer = `you have ${activities.length} todos - ${activities.join(
+      " and "
+    )}`;
+  }
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/homework.js:40]: I would consider making this into it's own function as you have separated all the others

File: javascript/javascript1/week4/homework.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    answer = add(num1, num2);
+  } else if (operation === "-") {
+    answer = subtract(num1, num2);
+  } else if (operation === "*") {
+    answer = multiply(num1, num2);
+  } else if (operation === "/") {
+    answer = divide(num1, num2);
+  } else {
+    answer = "Invalid operation";
+  }
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/homework.js]: This works and again great how you abstracted the functions away! To shorten it I would try something like this and call the function directly depending on the operator.

```
const operations = {
"+": add(num1,num2),
"-": subtract(num1,num2),
"*": multiply(num1,num2),
"/": divide(num1,num2),
};

// Pseudo:
// operations[operator]
```

File: javascript/javascript1/week4/homework.js [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      answer = `the ${activity} not found`;
+    }
+  }
+  // }
+  return answer;
+}
+
+//DATE
+function getDatOfToday() {
+  let msg = "";
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/homework.js]: No need to define it up here

File: javascript/javascript1/week4/homework.js [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+  const monthOfToday = today.getMonth();
+  const month = (f) =>
+    Array.from(Array(12), (e, i) =>
+      new Date(25e8 * ++i).toLocaleString("en-US", { month: f })
+    );
+
+  const day = today.getDate();
+
+  msg = `Today is: ${day}. of ${month`long`[monthOfToday]} ${year}`;
+  return msg;
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/homework.js]:
```
const msg = ....
return msg
```

File: javascript/javascript1/week4/homework.js [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+  const today = new Date();
+
+  const year = today.getFullYear();
+  //get month long alone
+  const monthOfToday = today.getMonth();
+  const month = (f) =>
+    Array.from(Array(12), (e, i) =>
+      new Date(25e8 * ++i).toLocaleString("en-US", { month: f })
+    );
```
Comment (islanderupnorth) [inline javascript/javascript1/week4/homework.js:72]: I'm not quite sure what you are attempting here but if it is to get today's month written out?

Then you could do as follows:

```
const today = new Date();
const monthIndex = today.getMonth(); // Returns the month (0-11)
const monthNames = [
  "January", "February", "March", "April", "May", "June",
  "July", "August", "September", "October", "November", "December"
];

const monthName = monthNames[monthIndex];
```

**Review by islanderupnorth — COMMENTED** [review]
Good work Alireza. You have understood the logic and you have a really good habit of abstracting the functions outside. This makes the `getReply()` really easy to read and understand what is happening! I would suggest you take the time to see how this exercise could be done using RegEx.

An example of how the checks could be done are as so:

```
const patterns = {
  namePattern: /Hello, my name is (.+)/i,
  addPattern: /Add (.+) to my to do list\./i,
  removePattern: /Remove (.+) from my to do list\./i,
  showToDoPattern: /What is my to do\?/,
  mathPattern: /what is (\d+)\s*([+\-*\/])\s*(\d+)\?/i,
  getNamePattern: /What is my name\?/i,
  datePattern: /What day is it today\?/i,
  timerPattern: /Set a timer for (\d+) min/i,
};
```

Then .test to see if it matches it in the conditionals.

You can [read more about regex here](https://www.w3schools.com/jsref/jsref_obj_regexp.asp)

It will require you Google each individual regex pattern but it comes in handy in the future!

File: javascript/javascript1/week4/exercise.js (hunk 1 above)
Comment (AliRezaHassanloo2731) [reply javascript/javascript1/week4/exercise.js]: it works, I checked again. the problem is that when we run it with big number such as 1000, it goes stack overflow

File: javascript/javascript1/week4/homework.js (hunk 4 above)
Comment (AliRezaHassanloo2731) [reply javascript/javascript1/week4/homework.js]: your are right, i create object for other reason, and then i didn't use it at all and forgot to clean it.

File: javascript/javascript1/week4/homework.js (hunk 8 above)
Comment (AliRezaHassanloo2731) [reply javascript/javascript1/week4/homework.js]: Done

**Comment by AliRezaHassanloo2731** [comment]
tanks for feedback

**Pull request merged**